RFC: Tabster for convergence/v9 (#17129)
microsoft/fluentui: rfcs/convergence/react-focus-management.md

# RFC: Introducing ability-helpers as @fluentui/react-focus-management

<!--
An RFC can be anything. A question, a suggestion, a plan. The purpose of this template is to give some structure to help folks write successful RFCs. However, don't feel constrained by this template; use your best judgement.

Tips for writing a successful RFC:

- Simple plain words that make your point, fancy words obfuscate
- Try to stay concise, but don't gloss over important details
- Try to write a neutral problem statement, not one that motivates your desired solution
- Remember, "Writing is thinking". It's natural to realize new ideas while writing your proposal
-->

---

> _List contributors to the proposal here_

Reviewer: ```suggestion
@ling1726
```

## Summary

> This RFC proposes the introduction of the [ability-helpers](https://github.com/microsoft/ability-helpers) library in Fluent as a focus management tool. The library will be wrapped in Fluent by a new package `@fluentui/react-focus-managent` which will expose a Context and hooks based API.

Reviewer: ```suggestion
This RFC proposes the introduction of the [ability-helpers](https://github.com/microsoft/ability-helpers) library in Fluent as a focus management tool. The library will be wrapped in Fluent by a new package `@fluentui/react-focus-managent` which will expose a React Context and hooks based API.
```

_nit_

> This initial RFC proposes a simple initial API for keyboard arrow navigation and focus traversing utilities. These APIs already have a testing ground in the `Menu` and `Accordion` components that are under development that rely on specific keyboard focus behaviours.

Reviewer: ```suggestion
This initial RFC proposes a simple initial API for keyboard arrow navigation and focus traversing utilities. These APIs already have a testing ground in the `Menu` (#) and `Accordion` (#) components that are under development that rely on specific keyboard focus behaviours.
```

Can we link PRs there?

<!-- If there is relevant background include it here -->

## Problem statement

<!--
Why are we making this change? What problem are we solving? What do we expect to gain from this?

This section is important as the motivation or problem statement is indepenent from the proposed change. Even if this RFC is not accepted this Motivation can be used for alternative solutions.

In the end, please make sure to present a neutral Problem statement, rather than one that motivates a particular solution
-->

> The current `FocusZone` and `FocusTrapZone` are React components that are responsible for keyboarding and focus trap behaviours in v7 and v0. The main difference between the two libraries is that v7 makes these components publicly available while v0 does not.

Reviewer: ```suggestion
```

I don't think that this is true, we are re-exporting them in `@fluentui/react-northstar`.

Author: Interesting, I don't  see any mention of them in the docsite though.. it's not too relevant so i will remove

Reviewer: https://fluentsite.z22.web.core.windows.net/0.52.0/accessibility
https://fluentsite.z22.web.core.windows.net/0.52.0/focus-zone
Not in the most obvious places, but docs are there 😏

The React component pattern for focus management is not ideal, since it breaks all React paradigms for component isolation. These components rely on complex DOM operations and event listeners to implement behaviour. The components also expose a class-based interface for focus operations on children elements, which further breaks this paradigm.

The current `FocusZone` components apply focus behaviours on **rendered DOM** as opposed to **visible DOM**. The Teams messaging stream is an example of what can happen if focus lands from rendered items, it can be an issue virtualization and may cause unnecessary scrolling which will:

- Force rerender items that aren't visible
- Cause unnecessary network requests due to force rerendering
- Extra narration for screen reader users with unnecessary items not rendered

Nested focusables are a source of issues on the current `FocusZone` component, where nested focusables behave inconsistently or incorrectly. Here are some recent issues related to this

> - [#8551](https://github.com/microsoft/fluentui/issues/8551)
> - [#12177](https://github.com/microsoft/fluentui/issues/12177)
> - [#16037](https://github.com/microsoft/fluentui/issues/16037)
> - [#13210](https://github.com/microsoft/fluentui/issues/13210)

Reviewer: ```suggestion
- [microsoft/fluentui#8551](https://github.com/microsoft/fluentui/issues/8551)
- [microsoft/fluentui#12177](https://github.com/microsoft/fluentui/issues/12177)
- [microsoft/fluentui#16037](https://github.com/microsoft/fluentui/issues/16037)
- [microsoft/fluentui#13210](https://github.com/microsoft/fluentui/issues/13210)
```

_nit_

## Detailed Design or Proposal

### Ability helpers

<!-- This is the bulk of the RFC. Explain the proposal or design in enough detail for the inteded audience to understand. -->

The `@fluentu/react-focus-management` package will act as a facade wrapper around [ability-helpers](https://github.com/microsoft/ability-helpers) and exposes a set of hooks and utilities for focus management. The [ability-helpers](https://github.com/microsoft/ability-helpers) library is designed to be a lower level utility that manages the state of focusable elements than any rendering framework.

[ability-helpers](https://github.com/microsoft/ability-helpers) manages focus on a lower level than the rendering framework and acts on the **visible** DOM. Since operations are run on visible DOM there is a much clearer path for handling dynamically loaded content that would otherwise require effects outside the rendering lifecycle in React components which breaks component isolation principles.

> Patterns in [ability-helpers](https://github.com/microsoft/ability-helpers) are **declarative**, and therefore decoupled more easily with framework and application code. Groups of focusables are managed by declaring `data-*` attributes which the lower level utilties use to define focusable behaviour.

Reviewer: I think that we should mention that users will tweak behaviors via options for hooks rather than tweaking `data-*` attributes.

Author: That's a good point, although I'm describing how AH works here, I would put that later on in the proposal section

Nesting focusables is a problem that ability-helpers explicitly tries to solve, bringing focus management to a lower level than the rendering framework tries to accomplish this consistently independent of rendering cycles.

### Initial API

#### FocusManagementProvider

Context provider that hosts the [ability-helpers](https://github.com/microsoft/ability-helpers) instance and its API objects.

> The below table defines what is included in the context value

Reviewer: ```suggestion
The below table defines what is included in the context value.
```

> Note that this table will increase as more ability-helpers features are included in Fluent

| Name      | Value        | Description                           |
| --------- | ------------ | ------------------------------------- |
| focusable | FocusableAPI | API for traversing focusable elements |

> The below table defines the props used by the provider component. All context values can be used as props in the provider component but are optional.

Reviewer: I don't think that it's a good idea to expose `internal__ahInstance` as a prop 🐱 Can you please add a code snippet to better understand difference between tables?

```tsx
interface AhContextValue {
  // something is there
}

const AhContext = React.createContext<>()

//

function App() {
  return <AhProvider {/* props that can be defined are there */} />
}
```

Reviewer 2: Agreed, exposing anything `internal` seems iffy to me.

| Deloser                | DeloserAPI         | Tools to help automatically restore focus when it gets lost without manually calling `.focus()` from application code |
| internal\_\_ahInstance | AbilityHelpersCore | The ability helpers instance, not expected to be be used directly                                                     |

#### Hooks

This spec proposes an API based on React hooks pattern. [ability-helpers](https://github.com/microsoft/ability-helpers) works mainly on adding `data-*` attributes to DOM elements to apply configurable behaviours to the 'groups' of focusables.

However some internal behaviours such as `modalizer` or `deloser` require DOM elements during the configuration phase, which will be most easily configured in React using `refs`.

Therefore it must be noted that **although some hooks might not require ref interaction and can simply return attributes to be spread, in an effort to maintain consistency refs are used, which fits the Fluent component design pattern which enforces ref forwarding**

```typescript
// Basic up/down keyboard navigation

> const ref = useKeyboardNavigationGroup();

Reviewer: Do I correctly get that a hook will apply `data-*` attributes directly on DOM and will overwrite user's input?

```tsx
function useKeyboardNavigationGroup() {
  const ref = React.useRef();

  React.useRef(() => {
    const el = ref.current;

    el.dataset.someDataAttr = "someValue";
    // etc.
  });

  return ref;
}
```

Author: Yes, the motivation behind it is interesting....

AH has 'basic' properties which can be applied to the DOM as data-* attributes (there is a specific prefix for AH attributes just can't recall what is is exactly). In an idea scenario we shouldn't need refs at all...

but there APIs such as focus delosing which uses `onFocusLost` callback which is a non-serializable extended prop that needs a DOM node to apply correctly. Hence the current decision to use `refs` but I am open to change after some initial experiences

Author: No more refs for now, we'll deal with extended props when it gets to that point

<div ref={ref}>{props.children}</div>;
```

```typescript

> const ref = React.useReact<HTMLElement>();

Reviewer: ```suggestion
const ref = React.useRef<HTMLElement>();
```

const {
  findFirst,
  findLast,
  findAll,

> } = useFocusable(ref); // All operations are with respect to the given DOM container

Reviewer: We should avoid APIs that are accepting refs:

```tsx
function App() {
  const ref = React.useRef();
  const [el, setEl] = React.useState();

  // 🟡 will work, but how we will know that `ref` changed?
  // https://github.com/theKashey/use-callback-ref tries to solve this issue, BTW
  useFocusable(ref);

  // 🔴 `el` will become something only after a render cycle, should we re-render? or do not support this case?
  useFocusable(el);

  return (
    <>
      <div ref={ref} /> /* both will work on React components */
      <div ref={setEl} />
    </>
  );
}
```

Author: Ah sorry the intention was to accept a DOM element not a ref

Reviewer: In this case, an issue related to DOM updates is still there:

```tsx
<MenuItem as={smth ? 'span' : 'div'} /> // 💣 `as` can be a different DOM node
```

In this case, DOM element might change and as we don't control user's input, we will not know about this change. This might not be an issue for this hook, but it's about a pattern in general.

Author: updated to use `ref.current` so a DOM element is accepted

Author: no more refs for now

findFirst() // first focusable
findLast() // first last focusable
findAll(condition: () => true) // first all focusables with condition

<div ref={ref}>{props.children}</div>
```

#### useFocusable

A hook that exposes helper methods to traverse focusable elements in the context of a DOM element.

The below table lists the helper functions to support:

| Name      | Value         | Options                  | Description                                         |
| --------- | ------------- | ------------------------ | --------------------------------------------------- |
| findFirst | HTMLElement   |                          | Finds the first focusable element                   |

> | findLast  | HTMLElement   |                          | Finds the last focusable element                    |

Reviewer: Just wondering. Will we have `findPrevious` and `findNext`?

Author: AH does expose them, but I'm not sure of a use case yet where they will be necessary, so I would rather iterate and add them when we have a use case for it

| findAll   | HTMLElement[] | boolean matcher function | Finds all focusable elements that match a condition |

> #### useKeyboardNagiationGroup

Reviewer: ```suggestion
#### useKeyboardNavigationGroup
```

A hook that returns a `ref` to be assigned to the container of a group of focusable elements that should support keyboard navigation

The below table lists the configuration options that the hook should support:

| Name          | Value                  | Description                                                                                               |
| ------------- | ---------------------- | --------------------------------------------------------------------------------------------------------- |
| axis          | vertical \| horizontal | Indicates which direction keyboard navigation should use                                                  |
| enableHomeEnd | boolean                | Enables the use of `Home` and `End` keys to navigate to first and last elements of the group respectively |

> | cyclic        | boolean                | Indicates that navigation should not stop at first and last elements respectively                         |

Reviewer: FYI: `FocusZone` uses `isCircularNavigation`, so may be `circular` is also an option there...

Author: updated

### Pros and Cons

<!-- Enumerate the pros and cons of the proposal. Make sure to think about and be clear on the cons or drawbacks of this propsoal. If there are multiple proposals include this for each. -->

#### Pros

> The current development pace of converged components means we will be able iterate quickly on this API since components under development such as `Menu`, `Accordion` and `Tooltip` will require these behaviours in the coming project cycles

Reviewer: ```suggestion
The current development pace of converged components means we will be able iterate quickly on this API since components under development such as `Menu`, `Accordion` and `Tooltip` will require these behaviours in the coming project cycles.
```

If `Tooltip` is mentioned, do we have a proposal for focus traps?

Author: I'm currently working on one, I can try add it to this RFC we'll see how it goes

The current `FocusZone` components apply focus behaviours on **rendered DOM** as opposed to **visible DOM**. The Teams messaging stream is an example of what can happen if focus lands from rendered items, it can be an issue virtualization and may cause unnecessary scrolling which will:

- Force rerender items that aren't visible
- Cause unnecessary network requests due to force rerendering
- Extra narration for screen reader users with unnecessary items not rendered

Nested focusables are a source of issues on the current `FocusZone` component, where nested focusables behave inconsistently or incorrectly. Here are some recent issues related to this

- [microsoft/fluentui#8551](https://github.com/microsoft/fluentui/issues/8551)
- [microsoft/fluentui#12177](https://github.com/microsoft/fluentui/issues/12177)
- [microsoft/fluentui#16037](https://github.com/microsoft/fluentui/issues/16037)
- [microsoft/fluentui#13210](https://github.com/microsoft/fluentui/issues/13210)

## Detailed Design or Proposal

### Ability helpers

<!-- This is the bulk of the RFC. Explain the proposal or design in enough detail for the inteded audience to understand. -->

> The `@fluentu/react-focus-management` package will act as a facade wrapper around [ability-helpers](https://github.com/microsoft/ability-helpers) and exposes a set of hooks and utilities for focus management. The [ability-helpers](https://github.com/microsoft/ability-helpers) library is designed to be a lower level utility that manages the state of focusable elements than any rendering framework.

Reviewer 2: nit:
```suggestion
The `@fluentui/react-tabster` package will act as a facade wrapper around [tabster](https://github.com/microsoft/tabster) and exposes a set of hooks and utilities for focus management. The [tabster](https://github.com/microsoft/tabster) library is designed to be a lower level utility that manages the state of focusable elements rather than relying on any rendering framework.
```

> [ability-helpers](https://github.com/microsoft/ability-helpers) manages focus on a lower level than the rendering framework and acts on the **visible** DOM. Since operations are run on visible DOM there is a much clearer path for handling dynamically loaded content that would otherwise require effects outside the rendering lifecycle in React components which breaks component isolation principles.

Reviewer 2: nit:
```suggestion
[tabster](https://github.com/microsoft/tabster) manages focus on a lower level than the rendering framework and acts on the **visible** DOM. Since operations are run on visible DOM there is a much clearer path for handling dynamically loaded content that would otherwise require effects outside the rendering lifecycle in React components which breaks component isolation principles.
```

> Patterns in [ability-helpers](https://github.com/microsoft/ability-helpers) are **declarative**, and therefore decoupled more easily with framework and application code. Groups of focusables are managed by declaring `data-*` attributes which the lower level utilties use to define focusable behaviour.

Reviewer 2: nit:
```suggestion
Patterns in [tabster](https://github.com/microsoft/tabster) are **declarative**, and therefore decoupled more easily with framework and application code. Groups of focusables are managed by declaring `data-*` attributes which the lower level utilities use to define focusable behaviour.
```

> Nesting focusables is a problem that ability-helpers explicitly tries to solve, bringing focus management to a lower level than the rendering framework tries to accomplish this consistently independent of rendering cycles.

Reviewer 2: nit:
```suggestion
Nesting focusables is a problem that `tabster` explicitly tries to solve, bringing focus management to a lower level than the rendering framework tries to accomplish this consistently, independent of rendering cycles.
```

### Modules

> The description of the each constituent module in the library can be found in the [ability-helpers README](https://github.com/microsoft/ability-helpers).

Reviewer 2: nit:
```suggestion
The description of the each constituent module in the library can be found in the [tabster README](https://github.com/microsoft/tabster).
```

The bundle size of each of the differnt modules of the library are listed in the below table. The library is fully tree shakeable, and for most cases only the `Core` part of the library is expected to be used

| Module          | minified (kb) | gzipped(kb |
|-----------------|---------------|------------|
| Core            | 45.3          | 12.2       |
| Deloser         | 13.2          | 3.2        |
| Modalizer       | 4.5           | 0.8        |
| ObservedElement | 3.1           | 0.7        |
| Outline         | 7.9           | 2          |
| CrossOrigin     | 18.9          | 4.4        |

### Initial API

#### FocusManagementProvider

> Context provider that hosts the [ability-helpers](https://github.com/microsoft/ability-helpers) instance and its API objects.

Reviewer 2: ```suggestion
Context provider that hosts the [tabster](https://github.com/microsoft/tabster) instance and its API objects.
```

The below table defines what is included in the context value.

> > Note that this table will increase as more ability-helpers features are included in Fluent

Reviewer 2: nit:
```suggestion
> Note that this table will increase as more `tabster` features are included in FluentUI
```

| Name      | Value        | Description                           |
| --------- | ------------ | ------------------------------------- |
| focusable | FocusableAPI | API for traversing focusable elements |

The below table defines the props used by the provider component. All context values can be used as props in the provider component but are optional.

| Name                   | Value              | Description                                                                                                           |
| ---------------------- | ------------------ | --------------------------------------------------------------------------------------------------------------------- |
| dir                    | 'ltr' \| 'rtl'     | Direction indicator, some keyboard behaviour will need to change based on this value                                  |
| Deloser                | DeloserAPI         | Tools to help automatically restore focus when it gets lost without manually calling `.focus()` from application code |

> | internal\_\_ahInstance | AbilityHelpersCore | The ability helpers instance, not expected to be be used directly                                                     |

Reviewer 2: > not expected to be used directly.

I think having this in the description is precisely the reason why we might want to not expose this as we've had trouble in the past with people doing/using things "not expected" to be done/used.

#### Hooks

> This spec proposes an API based on React hooks pattern. [ability-helpers](https://github.com/microsoft/ability-helpers) works mainly on adding `data-*` attributes to DOM elements to apply configurable behaviours to the 'groups' of focusables.

Reviewer 2: nit:
```suggestion
This spec proposes an API based on React hooks pattern. [tabster](https://github.com/microsoft/ability-helpers) works mainly by adding `data-*` attributes to DOM elements to apply configurable behaviours to the 'groups' of focusables.
```

However some internal behaviours such as `modalizer` or `deloser` require DOM elements during the configuration phase, which will be most easily configured in React using `refs`. However the initial implementation will be based on simple spreading of data attributes.

```typescript
// Basic up/down keyboard navigation

> const [attributes] = useKeyboardNavigationGroup();

Reviewer 2: Should this be:

```suggestion
const attributes = useKeyboardNavigationGroup();
```

The below table lists the helper functions to support:

| Name      | Value         | Options                  | Description                                         |
| --------- | ------------- | ------------------------ | --------------------------------------------------- |
| findFirst | HTMLElement   |                          | Finds the first focusable element                   |
| findLast  | HTMLElement   |                          | Finds the last focusable element                    |
| findAll   | HTMLElement[] | boolean matcher function | Finds all focusable elements that match a condition |

#### useKeyboardNavigationGroup

A hook that returns a set of data-* attributes to be spread to the container of a group of focusable elements that should support keyboard navigation

The below table lists the configuration options that the hook should support:

| Name          | Value                  | Description                                                                                               |
| ------------- | ---------------------- | --------------------------------------------------------------------------------------------------------- |

> | axis          | vertical \| horizontal | Indicates which direction keyboard navigation should use                                                  |

Reviewer 2: We should think about supporting `bidirectional` and `domOrder` here as well.

| enableHomeEnd | boolean                | Enables the use of `Home` and `End` keys to navigate to first and last elements of the group respectively |
| circular      | boolean                | Indicates that navigation should not stop at first and last elements respectively                         |

### Pros and Cons

<!-- Enumerate the pros and cons of the proposal. Make sure to think about and be clear on the cons or drawbacks of this propsoal. If there are multiple proposals include this for each. -->

#### Pros

The current development pace of converged components means we will be able iterate quickly on this API since components under development such as `Menu`, `Accordion` and `Tooltip` will require these behaviours in the coming project cycles

> Internally ability-helpers leverages the [TreeWalker](https://developer.mozilla.org/en-US/docs/Web/API/TreeWalker), which is the recommended the DOM tree traversal method [according to the w3c](https://www.w3.org/TR/DOM-Level-2-Traversal-Range/traversal.html). This API is dynamic and operations are designed represent underlying changes to the DOM which is a big plus for handling dynamic or lazy loaded content through a rendering framework.

Reviewer 2: nit:
```suggestion
Internally, `tabster` leverages [TreeWalker](https://developer.mozilla.org/en-US/docs/Web/API/TreeWalker), which is the recommended DOM tree traversal method [according to the w3c](https://www.w3.org/TR/DOM-Level-2-Traversal-Range/traversal.html). This API is dynamic and operations are designed to represent the underlying changes to the DOM, which is a big plus for handling dynamic or lazy loaded content through a rendering framework.
```

> A hook based implementation will be a lot easier to maintain for end users, since function calls can be added/removed/mocked easily which leads to easier adaptation in large codebases where extensive unit or snapshot testing is involved.

Reviewer 2: This is such a big plus for this approach! 👍

#### Cons

The library is new, and currently partially used in TMP. Not all issues have been identified. And contributions were still needed to the library to support the initial keyboard navigation API proposed in this RFC

> For backwards compat and support, the `FocusZone` components will still be needed for partners that don't use Function components in React, which are still relatively new.

Reviewer 2: ```suggestion
This is something interesting, is there a way we can abstract a new `FocusZone` component with the use of the `@fluentui/react-tabster` package to alleviate this transition? We could mark it as deprecated from the get-go.
```

> The complex bi-directional navigation `FocusZone` supports has not been explored with ability-helpers, and might require contributions back to the library to support such a case.

Reviewer 2: +1 on this, we should really strive to bring contributions back to `tabster` that add this functionality.

> The complex bi-directional navigation `FocusZone` supports has not been explored with ability-helpers, and might require contributions back to the library to support such a case.

Reviewer 2: nit:
```suggestion
The complex bi-directional navigation `FocusZone` supports has not been explored with `tabster`, and might require contributions back to the library to support such a case.
```
